File: Star_Discrepancy/QMC/Bundschuh_Zhu.py
import bisect

def Bundschuh_Zhu_Algorithm(pointset : list[list]) -> float:
    n = len(pointset)
    # Sorts pointset based on x values
    sorted_pointset = sorted(pointset, key=lambda coord : coord[0])
    # Extracts x and y values
    x_values = [0.0] + [entry[0] for entry in sorted_pointset] + [1.0]
    y_values = [entry[1] for entry in sorted_pointset]
    y_matrix_values = []
    max_discrepancy = 0
    for l in range(0, n+1):
        # Using a if statement and doing it by inserting a value is still faster than re-sorting
        if l > 0: bisect.insort(y_matrix_values, y_values[l-1])
        matrix_y = [0.0] + y_matrix_values + [1.0]
        for k in range(0, l+1):
            # Calculation from the formula
            point_ratio = k / n
            area_one = x_values[l] * matrix_y[k]
            area_two = x_values[l+1] * matrix_y[k+1]
            discrepancy = max(point_ratio - area_one, area_two - point_ratio)
            # If discrepancy is larger append it
            if(discrepancy > max_discrepancy):
                max_discrepancy = discrepancy
    return round(max_discrepancy, 7)

def Tovstik_Improvement(pointset : list[list]):
    N = len(pointset)
    
    sorted_pointset = sorted(pointset, key=lambda coord : coord[0])
    # Only seems to work without [0.0] and [1.0]
    x_values = [entry[0] for entry in sorted_pointset]
    y_values = [entry[1] for entry in sorted_pointset]
    
    # Adding max here also returns the incorrect value somehow
    d2 = 0
    
    for j in range(N):
        mu = 0  
        v = 0 
        for i in range(N):
            if y_values[i] <= y_values[j]:
                v += 1
                point_ratio = v / N
                area = x_values[i] * y_values[j]
                discrepancy_value = abs(point_ratio - area)
                d2 = max(d2, discrepancy_value)
                
                if mu > 1:
                    previous_point_ratio = (v - 1) / N
                    previous_discrepancy_value = abs(previous_point_ratio - area)
                    d2 = max(d2, previous_discrepancy_value)
            else:
                mu += 1  
    
    return round(d2, 7) 
 

File: src/c/Bundschuh_Zhu.c
typedef struct Point{
    float x, y;
} Point;

int compare (const void * a, const void * b) {
    float fa = ((Point*)a)->x;
    float fb = ((Point*)b)->x;
    return (fa > fb) - (fa < fb);
}

Point *test_pointset(){
    Point *pointset = (Point*)malloc(32*sizeof(Point));
    for(int k = 0; k < 33; k++){
        pointset[k].x = k / 32.0;
        pointset[k].y = fmod(7.0 * k / 32.0, 1);
    }
    return pointset;
}

float Bundschuh_Zhu_Algorithm(Point *pointset, int n) {
    qsort(pointset, n, sizeof(Point), compare);

    float *x_values = (float *)malloc((n + 2) * sizeof(float));
    float *y_values = (float *)malloc(n * sizeof(float));
    float *y_matrix_values = (float *)malloc(n * sizeof(float));
    
    int y_matrix_size = 0;
    float max_discrepancy = 0.0;

    x_values[0] = 0.0;
    x_values[n + 1] = 1.0;

    for (int i = 0; i < n; i++) {
        x_values[i + 1] = pointset[i].x;
        y_values[i] = pointset[i].y;
    }

    for (int l = 0; l <= n; l++) {
        // Insertion sort
        if (l > 0) {
            int i = y_matrix_size;
            while (i > 0 && y_matrix_values[i - 1] > y_values[l - 1]) {
                y_matrix_values[i] = y_matrix_values[i - 1];
                i--;
            }
            y_matrix_values[i] = y_values[l - 1];
            y_matrix_size++;
        }

        float *matrix_y = (float *)malloc((y_matrix_size + 2) * sizeof(float));
        matrix_y[0] = 0.0;
        for (int i = 0; i < y_matrix_size; i++) {
            matrix_y[i + 1] = y_matrix_values[i];
        }
        matrix_y[y_matrix_size + 1] = 1.0;

        for (int k = 0; k <= l; k++) {
            float point_ratio = (float)k / n;
            float area_one = x_values[l] * matrix_y[k];
            float area_two = x_values[l + 1] * matrix_y[k + 1];
            float discrepancy = fmax(point_ratio - area_one, area_two - point_ratio);
            if (discrepancy > max_discrepancy) {
                max_discrepancy = discrepancy;
            }
        }
        free(matrix_y);
    }

    free(x_values);
    free(y_values);
    free(y_matrix_values);

    return round(max_discrepancy * 1e7) / 1e7;
}